Cross-worktree cache reuse emits suppressed findings + "no such file or directory" warnings

### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Brew

### Description of the problem

**Summary**

If you run golangci-lint in two checkouts of the same commit that share a cache, findings you've suppressed can come back in the other checkout, pointing at a file that isn't there anymore. The default GOLANGCI_LINT_CACHE is machine-wide, so any setup with git worktrees hits this without doing anything special: create a new codegen file in a worktree, lint, merge to main, go back to main folder and pull, delete the worktree, lint, boom. You see it as findings you thought were suppressed (whether by //nolint, generated-file detection, or path exclusions), usually alongside no such file or directory warnings. Clearing the cache makes it go away until it happens again.

**Root Cause**

The cache is keyed by file content rather than path, and it's shared across the machine, so two checkouts of the same commit end up sharing entries. The problem is what's stored in the entry: a cached issue keeps the absolute Pos.Filename from whichever checkout produced it. Several processors decide suppression by re-reading the source on every run (nolint_filter, generated_file_filter, source_code), and they read that stored path. If the checkout it came from is gone, the read fails, the processor treats the issue as not-suppressed, and it gets reported against a path that no longer exists.

**Proposed Fix**

Store the position relative to the module root in the cache entry, then turn it back into an absolute path against the current checkout's module root when the entry is loaded. This is safe because the cache key already normalizes in-module paths to module-relative (internal/cache/cache.go), so a cache hit means the two checkouts have the same file at the same relative location, and the local copy is a fine substitute. It covers all three processors, and it doesn't change how the cache is shared or how it performs, since only the stored value changes. Paths outside the current module (dependencies and the like) are left absolute.

Other options I looked at: store the file-derived facts (is-generated, //nolint ranges) in the cache next to the diagnostics, or fail closed when a cached issue's source can't be read.

**Related:** Probably the same underlying bug as #3502 and #6656. #3502 (open) reports the no such file or directory warning but reads it as a "clear your cache" annoyance, missing that findings actually get through. #6656 (closed) traced it to the absolute path for path exclusions but concluded worktrees just can't share a cache safely. Filing this because it's the same bug for //nolint, generated files, and source lines too, and proposes a fix that keeps the cache shared across checkouts.

### Version of golangci-lint

golangci-lint has version 2.12.2 built with go1.26.2 from c0d3ddc9 on 2026-05-06T11:07:58Z

### Configuration

<details>

Config file (.golangci.yml):
```console
version: "2"
linters:
  default: none
  enable: [ineffassign]
```
CLI flags (same invocation in both worktrees):
```console
golangci-lint run ./... --issues-exit-code 0
```

</details>

### Go environment

<details>

```
go version go1.26.5 darwin/arm64
AR='ar'
CC='clang'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='clang++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/daxfohl/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/daxfohl/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT='jsonv2'
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/go-build2138730563=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/dev/null'
GOMODCACHE='/Users/daxfohl/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/daxfohl/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/nix/store/naw4mj4s58ymzlc10a76y79m40w3wfpp-go-1.26.5/share/go'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/Users/daxfohl/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/nix/store/naw4mj4s58ymzlc10a76y79m40w3wfpp-go-1.26.5/share/go/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.26.5'
GOWORK='/Users/daxfohl/monorepo/go.work'
PKG_CONFIG='pkg-config'
```

</details>

### Verbose output of running

<details>

```console
INFO golangci-lint has version 2.12.2 built with go1.26.2 from c0d3ddc on 2026-05-06T11:01:25Z
INFO [config_reader] Config search paths: [./ /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.6c9ANCtVWl /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn /var/folders/5v /var/folders /var / /Users/daxfohl]
INFO [config_reader] Used config file .golangci.yml
INFO [config_reader] Module name "x"
INFO [goenv] Read go env for 8.229167ms: map[string]string{"GOCACHE":"/Users/daxfohl/Library/Caches/go-build", "GOROOT":"/opt/homebrew/Cellar/go/1.26.5/libexec"}
INFO [lintersdb] Active 1 linters: [ineffassign]
INFO [loader] Go packages loading at mode 8199 (name|compiled_files|files) took 26.713917ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 87.75µs
INFO [linters_context/goanalysis] analyzers took 63.125µs with top 10 stages: ineffassign: 57.75µs, typecheck: 5.375µs
INFO [runner] Issues before processing: 1, after processing: 0
INFO [runner] Processors filtering stat (in/out): exclusion_paths: 1/1, nolint_filter: 1/0, path_absoluter: 1/1, invalid_issue: 1/1, path_relativity: 1/1, cgo: 1/1, filename_unadjuster: 1/1, generated_file_filter: 1/1, exclusion_rules: 1/1
INFO [runner] processing took 58.167µs with stages: nolint_filter: 30.292µs, generated_file_filter: 22.583µs, path_relativity: 1.583µs, sort_results: 1.209µs, cgo: 375ns, invalid_issue: 333ns, max_same_issues: 291ns, filename_unadjuster: 250ns, fixer: 167ns, source_code: 166ns, path_shortener: 125ns, diff: 125ns, path_absoluter: 125ns, exclusion_paths: 125ns, max_per_file_from_linter: 84ns, exclusion_rules: 84ns, path_prettifier: 84ns, max_from_linter: 83ns, uniq_by_line: 83ns, severity-rules: 0s
INFO [runner] linters took 7.375084ms with stages: goanalysis_metalinter: 7.278375ms
0 issues.
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 2 samples, avg is 43.9MB, max is 43.9MB
INFO Execution took 42.6335ms
```

</details>

### A minimal reproducible example or link to a public repository

<details>

Set up two worktrees, A and B, from the same commit. The one finding in the file is suppressed with an inline //nolint. The trick is the order:

1. Lint B first, with an empty cache. It comes back clean, and golangci-lint caches that result, keyed by the file's contents, but storing B's absolute path.
2. Delete worktree B but keep the cache around. Now that stored path points at nothing.
3. Lint A. The file is byte-for-byte identical to B's, so it's a cache hit, and the cached result still refers to the file over in the deleted B.

```zsh
work="$(mktemp -d)"; cache="$(mktemp -d)"
export GOLANGCI_LINT_CACHE="$cache/cache"
cd "$work"

git init -q && git config user.email a@b.c && git config user.name r
printf 'module x\n\ngo 1.21\n' > go.mod
printf 'package x\n\nfunc F() int {\n\ty := 1 //nolint:ineffassign\n\ty = 2\n\treturn y\n}\n' > x.go
printf 'version: "2"\nlinters:\n  default: none\n  enable: [ineffassign]\n' > .golangci.yml
git add -A && git commit -qm init
git worktree add -q A HEAD && git worktree add -q B HEAD

# 1. Lint B: comes back clean, and caches the result under B's path.
( cd B && golangci-lint run ./... )

# 2. Delete B but keep the cache. That stored path now points at nothing.
git worktree remove --force B

# 3. Lint A: same file contents, so it's a cache hit — carrying B's deleted path.
( cd A && golangci-lint run ./... )

# Clean up.
cd /; rm -rf "$work" "$cache"; unset GOLANGCI_LINT_CACHE
```


**Expected**

Both runs should say 0 issues. It's the same file in both worktrees, and the //nolint on it suppresses the only finding.

**Actual**

B's run is clean, as it should be. But once B is gone, linting A hits the cache, picks up B's deleted path, and reports the finding anyway, along with two warnings from processors that tried to re-read the file that isn't there:

```sh
level=warning msg="[runner] Can't process results by generated_file_filter processor: can't filter issue &result.Issue{FromLinter:\"ineffassign\", Text:\"ineffectual assignment to y\", Severity:\"\", SourceLines:[]string(nil), Pkg:(*packages.Package)(0x7386b441f860), Pos:token.Position{Filename:\"/var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go\", Offset:27, Line:4, Column:2}, LineRange:(*result.Range)(nil), HunkPos:0, SuggestedFixes:[]analysis.SuggestedFix(nil), ExpectNoLint:false, ExpectedNoLintLinter:\"\", WorkingDirectoryRelativePath:\"../../../../../../../../var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go\", RelativePath:\"../B/x.go\"}: failed to get doc (strict) of file /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go: failed to parse file: open /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go: no such file or directory"
level=warning msg="[runner/source_code] Failed to get line 4 for file /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go: failed to get file /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go lines cache: can't get file /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go bytes from cache: can't read file /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go: open /var/folders/5v/fskgkrqx4_q57k6nwdglj1y00000gn/T/tmp.erlLsukndw/B/x.go: no such file or directory"
../B/x.go:4:2: ineffectual assignment to y (ineffassign)
1 issues:
* ineffassign: 1
```
Since it's a scratch cache and both temp dirs get removed at the end, nothing touches your real cache or lingers afterward. If you want to poke around before it's wiped, just don't paste that last line until you're done.
</details>

### Validation

- [x] Yes, I've included all the information above (example, version, config, etc.).

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.

--- a/internal/cache/cache.go
+++ b/internal/cache/cache.go
@@ -7,7 +7,6 @@ import (
 	"errors"
 	"fmt"
 	"maps"
-	"path/filepath"
 	"runtime"
 	"slices"
 	"strings"
@@ -171,11 +170,6 @@ func (c *Cache) computePkgHash(pkg *packages.Package) (hashResults, error) {
 		h, fErr := c.fileHash(f)
 		if fErr != nil {
 			return nil, fmt.Errorf("failed to calculate file %s hash: %w", f, fErr)
-		}
-
-		// This is the current module (the project to analyze).
-		if pkg.Module != nil && pkg.Module.Version == "" {
-			f = pkg.Module.Path + strings.TrimPrefix(filepath.ToSlash(f), filepath.ToSlash(pkg.Module.Dir))
 		}
 
 		fmt.Fprintf(key, "file %s %x\n", f, h)

--- a/internal/cache/cache_test.go
+++ b/internal/cache/cache_test.go
@@ -2,8 +2,6 @@ package cache
 
 import (
 	"fmt"
-	"os"
-	"path/filepath"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -157,30 +155,4 @@ func TestCache_computeHash(t *testing.T) {
 	assert.Equal(t, "8978e3d76c6f99e9663558d7147a7790f229a676804d1fde706a611898547b74", results[HashModeNeedOnlySelf])
 	assert.Equal(t, "b1aef902a0619b5cbfc2d6e2e91a73dd58dd448e58274b2d7a5ff8efd97aefa4", results[HashModeNeedDirectDeps])
 	assert.Equal(t, "9c602ef861197b6807e82c99caa7c4042eb03c1a92886303fb02893744355131", results[HashModeNeedAllDeps])
-}
-
-func TestCache_computeHash_module(t *testing.T) {
-	pkgCache := setupCache(t)
-
-	// This creates a new random directory for each test run, but the hash will stay consistent.
-	tempDir := t.TempDir()
-
-	err := os.WriteFile(filepath.Join(tempDir, "foo.go"), []byte("package foo"), 0600)
-	require.NoError(t, err)
-
-	pkg := fakePackage()
-	pkg.Module = &packages.Module{
-		Dir:  tempDir,
-		Path: "github.com/golangci/example",
-	}
-	pkg.IgnoredFiles = []string{filepath.Join(tempDir, "foo.go")}
-
-	results, err := pkgCache.computePkgHash(pkg)
-	require.NoError(t, err)
-
-	require.Len(t, results, 3)
-
-	assert.Equal(t, "ac5d79f4630d6b5f1e4ac88bfa9974698ea44ea1f53760bc75001bd7c9ce9064", results[HashModeNeedOnlySelf])
-	assert.Equal(t, "73fd7ef46b20efdd1bff7eaedade26b1655ee727d83f4210b7934bf4d4a1ac1d", results[HashModeNeedDirectDeps])
-	assert.Equal(t, "063ffbaa2ea6b2a9a02724effe2a616b88672a4d8c3c4c6c8d777bfe5a49e7b4", results[HashModeNeedAllDeps])
 }

--- a/pkg/commands/run.go
+++ b/pkg/commands/run.go
@@ -713,8 +713,7 @@ func computeGoModSalt() (string, error) {
 		return "", fmt.Errorf("failed to read go.mod: %w", err)
 	}
 
-	// NOTE: the variable `goModPath` is not used here to ensure getting the same hash, independently of the location, for the same content.
-	sum, err := dirhash.Hash1([]string{"go.mod"}, func(string) (io.ReadCloser, error) {
+	sum, err := dirhash.Hash1([]string{goModPath}, func(string) (io.ReadCloser, error) {
 		return io.NopCloser(bytes.NewReader(data)), nil
 	})
 	if err != nil {

